examples/energy-management-app/energy-management-common/src/EVSEManufacturerImpl.cpp

```diff
@@ -89,6 +93,182 @@ CHIP_ERROR EVSEManufacturer::Shutdown()
+}
+
+/**
+ * @brief   Allows a client application to send in energy readings into the system
+ *
+ *          This is a helper function to add timestamps to the readings
+ *
+ * @param[in]  aCumulativeEnergyImported -total energy imported in milli-watthours
+ * @param[in]  aCumulativeEnergyExported -total energy exported in milli-watthours
+ */
+CHIP_ERROR EVSEManufacturer::SendEnergyReading(EndpointId aEndpointId, int64_t aCumulativeEnergyImported,
+                                               int64_t aCumulativeEnergyExported)
+{
+    MeasurementData * data = MeasurementDataForEndpoint(aEndpointId);
+
+    EnergyMeasurementStruct::Type energyImported;
+    EnergyMeasurementStruct::Type energyExported;
+
+    uint32_t oldTimestamp;
+
+    // Get current timestamp
+    uint32_t currentTimestamp;
+    CHIP_ERROR err = GetEpochTS(currentTimestamp);
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(AppServer, "GetEpochTS returned error getting timestamp");
+        return err;
+    }
+
+    /** IMPORT */
+    // Copy last endTimestamp into new startTimestamp if it exists
+    energyImported.startTimestamp.ClearValue();
+    if (data->cumulativeImported.HasValue())
+    {
+        if (data->cumulativeImported.Value().endTimestamp.HasValue())
+        {
+            oldTimestamp = data->cumulativeImported.Value().endTimestamp.Value();
+            energyImported.startTimestamp.SetValue(oldTimestamp);
+        }
+    }
+
+    energyImported.endTimestamp.SetValue(currentTimestamp);
+    energyImported.energy = aCumulativeEnergyImported;
+
+    /** EXPORT */
+    // Copy last endTimestamp into new startTimestamp if it exists
+    energyExported.startTimestamp.ClearValue();
+    if (data->cumulativeExported.HasValue())
+    {
+        if (data->cumulativeExported.Value().endTimestamp.HasValue())
+        {
+            oldTimestamp = data->cumulativeExported.Value().endTimestamp.Value();
+            energyExported.startTimestamp.SetValue(oldTimestamp);
+        }
+    }
+    energyExported.endTimestamp.SetValue(currentTimestamp);
+    energyExported.energy = aCumulativeEnergyExported;
+
+    // call the SDK to update attributes and generate an event
+    NotifyCumulativeEnergyMeasured(aEndpointId, MakeOptional(energyImported), MakeOptional(energyExported));
+
+    return CHIP_NO_ERROR;
+}
+
+struct FakeReadingsData
+{
+    bool bEnabled;                    /* If enabled then the timer callback will re-trigger */
+    EndpointId mEndpointId;           /* Which endpoint the meter is on */
+    uint8_t mInterval_s;              /* Interval in seconds to callback */
+    int64_t mPower_mW;                /* Power on the load in mW (signed value) +ve = imported */
+    uint32_t mPowerRandomness_mW;     /* The amount to randomize the Power on the load in mW */
+    int64_t mTotalEnergyImported = 0; /* Energy Imported which is updated if mPower > 0 */
+    int64_t mTotalEnergyExported = 0; /* Energy Imported which is updated if mPower < 0 */
```

Comment:
if we go both ways, are these always positive? if so should they be of an unsigned type?

Reply:
They do only go positive, but the underlying energy types are always signed in the spec. So in order to avoid casting issues when sending out over the wire/air I kept them as signed here too.

Reply:
Added a comment to explain.